cmd/promtool/unittest.go

```diff
@@ -0,0 +1,467 @@
+	GroupEvalOrder     []string      `yaml:"group_eval_order"`
+	Tests              []testGroup   `yaml:"tests"`
+}
+
+func (utf *unitTestFile) maxEvalTime() time.Duration {
+	var maxd time.Duration
+	for _, t := range utf.Tests {
+		d := t.maxEvalTime()
+		if d > maxd {
+			maxd = d
+		}
+	}
+	return maxd
+}
+
+// testGroup is a group of input series and tests associated with it.
+type testGroup struct {
+	Interval        time.Duration    `yaml:"interval"`
+	InputSeries     []series         `yaml:"input_series"`
+	AlertRuleTests  []alertTestCase  `yaml:"alert_rule_test,omitempty"`
+	PromqlExprTests []promqlTestCase `yaml:"promql_expr_test,omitempty"`
+}
+
+// test performs the unit tests.
+func (tg *testGroup) test(mint, maxt time.Time, evalInterval time.Duration, groupOrderMap map[string]int, ruleFiles ...string) []error {
+	// Setup testing suite.
+	suite, err := promql.NewTest(nil, tg.seriesLoadingString())
+	if err != nil {
+		return []error{err}
+	}
+	defer suite.Close()
+
+	err = suite.Run()
+	if err != nil {
+		return []error{err}
+	}
+
+	// Load the rule files.
+	opts := &rules.ManagerOptions{
+		QueryFunc:  rules.EngineQueryFunc(suite.QueryEngine(), suite.Storage()),
+		Appendable: suite.Storage(),
+		Context:    context.Background(),
+		NotifyFunc: func(ctx context.Context, expr string, alerts ...*rules.Alert) {},
+	}
+	m := rules.NewManager(opts)
+	groupsMap, ers := m.LoadGroups(tg.Interval, ruleFiles...)
+	if ers != nil {
+		return ers
+	}
+	groups := orderedGroups(groupsMap, groupOrderMap)
+
+	// Pre-processing some data for testing alerts.
+	// All this preparating so that we can test the alerts as and when we evaluate
+	// the rules so that we can avoid storing them in memory, as number of eval might
+	// get very big.
+
+	// All the `eval_time` for which we have unit tests.
+	var alertEvalTimes []time.Duration
+	// Map of all the eval_time+alertname combination present in the unit tests.
+	alertsInTest := make(map[time.Duration]map[string]struct{})
+	// Map of all the unit tests for given eval_time.
+	alertTests := make(map[time.Duration][]alertTestCase)
+	for _, alert := range tg.AlertRuleTests {
+		alertEvalTimes = append(alertEvalTimes, alert.EvalTime)
+
+		if _, ok := alertsInTest[alert.EvalTime]; !ok {
+			alertsInTest[alert.EvalTime] = make(map[string]struct{})
+		}
+		alertsInTest[alert.EvalTime][alert.Alertname] = struct{}{}
+
+		alertTests[alert.EvalTime] = append(alertTests[alert.EvalTime], alert)
+	}
+	sort.Slice(alertEvalTimes, func(i, j int) bool {
+		return alertEvalTimes[i] < alertEvalTimes[j]
+	})
+
+	// Current index in alertEvalTimes what we are looking at.
+	curr := 0
+
+	var errs []error
+	for ts := mint; ts.Before(maxt); ts = ts.Add(evalInterval) {
+		// Collects the alerts asked in unit testing.
+		for _, g := range groups {
+			g.Eval(suite.Context(), ts)
+		}
+
+		for curr < len(alertEvalTimes) && ts.Sub(mint) <= alertEvalTimes[curr] && alertEvalTimes[curr] < ts.Add(evalInterval).Sub(mint) {
```

Comment:
This is a bit long of a for expression, I'd suggest moving it into an if inside the for